src/main/java/com/mraof/minestuck/network/EditmodeDragPacket.java

```diff
@@ -0,0 +1,281 @@
+package com.mraof.minestuck.network;
+
+import com.mraof.minestuck.MinestuckConfig;
+import com.mraof.minestuck.alchemy.*;
+import com.mraof.minestuck.computer.editmode.*;
+import com.mraof.minestuck.player.PlayerSavedData;
+import com.mraof.minestuck.skaianet.SburbConnection;
+import com.mraof.minestuck.util.MSCapabilities;
+import com.mraof.minestuck.util.MSSoundEvents;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.util.Tuple;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SoundType;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public final class EditmodeDragPacket
+{
+	private static final Logger LOGGER = LogManager.getLogger();
+	
+	private static Tuple<Boolean, GristSet> editModePlaceCheck(EditData data, Player player, InteractionHand hand, BlockPos pos)
+	{
+		SburbConnection connection = data.getConnection();
+		ItemStack stack = player.getMainHandItem();
+		
+		if(stack.isEmpty() || hand.equals(InteractionHand.OFF_HAND) || !player.level.getBlockState(pos).getMaterial().isReplaceable())
+			return new Tuple<>(false, new GristSet());
+
+		DeployEntry entry = DeployList.getEntryForItem(stack, connection, player.level);
+		GristSet cost = entry != null ? entry.getCurrentCost(connection) : GristCost.findCostForItem(stack, null, false, player.level);
+		if(!GristHelper.canAfford(PlayerSavedData.getData(connection.getClientIdentifier(), player.level).getGristCache(), cost))
+			return new Tuple<>(false, cost);
+		
+		if(entry == null && !(stack.getItem() instanceof BlockItem))
+			return new Tuple<>(false, new GristSet());
+		
+		return new Tuple<>(true, new GristSet());
+	}
+	
+	private static Tuple<Boolean, GristSet> editModeDestroyCheck(EditData data, Player player, BlockPos pos)
```

Comment:
The tuple might not be a great fit here. I recommend changing it to just return a boolean, and instead take in a `Consumer<GristSet> missingCostTracker` (that you create using `missingCost::addGrist`)